sql_generators/glean_usage/templates/baseline_clients_last_seen_v1.query.sql

```diff
@@ -70,7 +70,12 @@ FROM
   _current
 FULL JOIN
   _previous
-USING
-  (client_id)
+ON
+  _previous.client_id = _current.client_id AND
+  _previous.sample_id = _current.sample_id AND
+  (
+    _previous.normalized_channel = _current.normalized_channel OR
+    (_previous.normalized_channel IS NULL AND _current.normalized_channel IS NULL)
+  )
```

Comment:
The contract for this table should be that there's only one row per client_id per `submission_date`. Do we meet that guarantee using this modified join condition?


The baseline_clients_* ETL is per appId whereas metrics_clients_* is per-appName. I don't think this logic makes sense for per-appId logic. We should be able to guarantee only a single client_id in the daily table.

Reply:
Okay, so I have been digging a little more and it appears that the client daily table actually does have multiple entries for a single client per `submission_date`. The reason for this is that in the `baseline` table some clients  have baseline pings with a `sample_id` and sometimes these same clients send baseline pings with no `sample_id` (`sample_id = NULL`). For these clients `clients_daily` has two entries.

It looks like this might have been an ingestion issue which has been fixed (I'm only seeing this for clients in 2019).
After 2019 `clients_daily` has indeed a single entry per client.

`clients_last_seen` still have duplicate entries to this date however, since it always joins on itself for the previous date. I think the way to go is to filter out entries where `sample_id` is `NULL`.

Reply:
> It looks like this might have been an ingestion issue which has been fixed (I'm only seeing this for clients in 2019).

Indeed, we didn't add support for parsing client_id out of Glean pings to populate sample_id until May 2019. Implemented in https://github.com/mozilla/gcp-ingestion/pull/612

Reply:
> I think the way to go is to filter out entries where sample_id is NULL

That would be fine. Even better would be to add sample_id in to the stable tables where it's missing, which we could do via query (and the data volume should be pretty tiny for 2019 data).